File: app/src/main/java/com/olehmesh/asinfo2/MyProvider.java
package com.olehmesh.asinfo2;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.widget.RemoteViews;

import java.util.Arrays;

public class MyProvider extends AppWidgetProvider {

    final String LOG_TAG = "myLogs";
    final static String ITEM_POSITION = "item_position";
    final String ONCLICK_ACTION = "onclick";

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager,
                         int[] appWidgetIds) {
        super.onUpdate(context, appWidgetManager, appWidgetIds);

            RemoteViews remoteViews = new RemoteViews(context.getPackageName(),
                    R.layout.widget);

            Intent intentAdapter = new Intent(context, MyService.class);
            intentAdapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds);
            remoteViews.setRemoteAdapter(R.id.listView, intentAdapter);

            setListClick(remoteViews, context);

            appWidgetManager.updateAppWidget(appWidgetIds, remoteViews);
            appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetIds,
                    R.id.listView);

        Log.d(LOG_TAG, "onUpdate " + Arrays.toString(appWidgetIds));

        }

    void setListClick(RemoteViews rv, Context context) {
        Intent listClickIntent = new Intent(context, MyProvider.class);
        listClickIntent.setAction(ONCLICK_ACTION);
        PendingIntent listClickPIntent = PendingIntent.getBroadcast(context, 0,
                listClickIntent, 0);
        rv.setPendingIntentTemplate(R.id.listView, listClickPIntent);
    }


    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);

        if (intent.getAction().equalsIgnoreCase(ONCLICK_ACTION)) {
            ComponentName thisAppWidget = new ComponentName(
                    context.getPackageName(), getClass().getName());
            AppWidgetManager appWidgetManager = AppWidgetManager
                    .getInstance(context);
            int ids[] = appWidgetManager.getAppWidgetIds(thisAppWidget);
            for (int appWidgetID : ids) {
                onUpdate(context, appWidgetManager, new int[]{appWidgetID});

            }
        }
    }

    }

File: app/src/main/java/com/olehmesh/asinfo2/MyProvider.kt
package com.olehmesh.asinfo2

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.util.Log
import android.widget.RemoteViews
import java.util.*

class MyProvider : AppWidgetProvider() {
    val LOG_TAG = "myLogs"
    val ONCLICK_ACTION = "onclick"
    override fun onUpdate(
        context: Context, appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)
        val remoteViews = RemoteViews(
            context.packageName,
            R.layout.widget
        )
        val intentAdapter = Intent(context, MyService::class.java)
        intentAdapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds)
        remoteViews.setRemoteAdapter(R.id.listView, intentAdapter)
        setListClick(remoteViews, context)
        appWidgetManager.updateAppWidget(appWidgetIds, remoteViews)
        appWidgetManager.notifyAppWidgetViewDataChanged(
            appWidgetIds,
            R.id.listView
        )
        Log.d(LOG_TAG, "onUpdate " + appWidgetIds.contentToString())
    }

    private fun setListClick(rv: RemoteViews, context: Context?) {
        val listClickIntent = Intent(context, MyProvider::class.java)
        listClickIntent.action = ONCLICK_ACTION
        val listClickPIntent = PendingIntent.getBroadcast(
            context, 0,
            listClickIntent, 0
        )
        rv.setPendingIntentTemplate(R.id.listView, listClickPIntent)
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        if (Objects.requireNonNull(intent.action).equals(ONCLICK_ACTION, ignoreCase = true)) {
            val thisAppWidget = ComponentName(
                context.packageName, javaClass.name
            )
            val appWidgetManager = AppWidgetManager
                .getInstance(context)
            val ids = appWidgetManager.getAppWidgetIds(thisAppWidget)
            for (appWidgetID in ids) {
                onUpdate(context, appWidgetManager, intArrayOf(appWidgetID))
            }
        }
    }

    companion object {
        const val ITEM_POSITION = "item_position"
    }
}